src/main/java/io/nats/client/impl/Headers.java

```diff
@@ -0,0 +1,274 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package io.nats.client.impl;
+
+import java.util.*;
+
+import static java.nio.charset.StandardCharsets.US_ASCII;
+
+public class Headers {
+	private static final String VERSION = "NATS/1.0";
+	private static final byte[] VERSION_BYTES = "NATS/1.0\r\n".getBytes(US_ASCII);
+	private static final byte[] COLON_BYTES = ":".getBytes(US_ASCII);
+	private static final byte[] CRLF_BYTES = "\r\n".getBytes(US_ASCII);
+	private static final int VERSION_BYTES_LEN = VERSION_BYTES.length;
+	private static final int COLON_BYTES_LEN = COLON_BYTES.length;
+	private static final int CRLF_BYTES_LEN = CRLF_BYTES.length;
+
+	private static final String KEY_CANNOT_BE_EMPTY_OR_NULL = "Header key cannot be null.";
+	private static final String KEY_INVALID_CHARACTER = "Header key has invalid character: ";
+	private static final String VALUE_INVALID_CHARACTERS = "Header value has invalid character: ";
+
+	private final Map<String, List<String>> headerMap;
+	private byte[] serialized;
+
+	public Headers() {
+		headerMap = new HashMap<>();
+	}
+
+	public Headers(byte[] serialized) {
+		this();
+		if (serialized != null && serialized.length > 0) {
+			String[] split = new String(serialized, US_ASCII).split("\\Q\r\n\\E");
+			if (!split[0].equals(VERSION)) {
+				throw new IllegalArgumentException("Invalid header version");
+			}
+			for (int x = 1; x < split.length; x++) {
+				String[] pair = split[x].split(":");
+				add(pair[0], pair[1].trim().split(","));
+			}
+		}
+	}
+
+	/**
+	 * If the key is present add the values to the list of values for the key.
+	 * If the key is not present, sets the specified values for the key.
+	 * null values are ignored. If all values are null, the key is not added or updated.
+	 *
+	 * @param key the key
+	 * @param values the values
+	 * @throws IllegalArgumentException if the key is null or empty or contains invalid characters
+	 *         -or- if any value contains invalid characters
+	 */
+	public Headers add(String key, String... values) {
+		if (values != null) {
+			_add(key, Arrays.asList(values));
+		}
+		return this;
+	}
+
+	/**
+	 * If the key is present add the values to the list of values for the key.
+	 * If the key is not present, sets the specified values for the key.
+	 * null values are ignored. If all values are null, the key is not added or updated.
+	 *
+	 * @param key the key
+	 * @throws IllegalArgumentException if the key is null or empty or contains invalid characters
+	 *         -or- if any value contains invalid characters
+	 */
+	public Headers add(String key, Collection<String> values) {
+		_add(key, values);
+		return this;
+	}
+
+	// the add delegate
+	private void _add(String key, Collection<String> values) {
+		if (values != null) {
```

Comment:
You should have this function convert all `String`'s to `ByteBuffer`'s before inserting into `headerMap`. Accessors that need to return strings should do the reverse.